certbot/main.py

```diff
@@ -0,0 +1,75 @@
+import datetime
+import os
+import subprocess
+import sys
+import time
+
+ACME_SERVER = "https://acme-v01.api.letsencrypt.org/directory"
+CERTBOT_DIRECOTRY = "/etc/letsencrypt/live"
+
+def run_certbot(header, args):
+  print("===============================================================================")
+  print(header)
+  print("===============================================================================")
+
+  sys.stdout.flush()
+  ec = subprocess.call(args)
+
+  print("certbot exit code " + str(ec))
+  print("")
+
+def run_certbot_certonly(domain, email):
+  run_certbot("Getting certificate for " + domain, ["certbot", "certonly", "--standalone",
+    "--agree-tos", "--noninteractive", "--text", "--server", ACME_SERVER, "--expand",
+    "--preferred-challenges", "http-01", "--email", email, "-d", domain])
+
+def run_certbot_renew():
+  post_hook = sys.executable + " " + sys.argv[0] + " post-hook"
+  run_certbot("Renewing certificates", ["certbot", "renew", "--post-hook", post_hook])
+
+def get_certificates():
+  email = os.getenv('CERTBOT_EMAIL', '')
+  for domain in os.environ['CERTBOT_DOMAINS'].split():
+    run_certbot_certonly(domain, email)
+
+def concat_certificates():
+  output_directory = os.getenv('CERTBOT_OUTPUT_DIRECTORY', '/certs')
+  if not os.path.exists(output_directory):
+    os.makedirs(output_directory)
+  for name in os.listdir(CERTBOT_DIRECOTRY):
+    path = os.path.join(CERTBOT_DIRECOTRY, name)
+    if os.path.isdir(path):
+      fp = open(os.path.join(output_directory, name + '.pem'), 'w')
+
+      for pem in ['fullchain.pem', 'privkey.pem']:
+        with open(os.path.join(path, pem), 'r') as fullchain:
+          fp.write(fullchain.read())
+          fp.write('\n')
+
+      fp.close()
+
+def touch_file():
+  path = os.getenv('CERTBOT_TOUCH_FILE')
+  if path:
+    with open(path, 'a'):
+      os.utime(path, None)
+
+def run_post():
+  concat_certificates()
+  touch_file()
+
+def run_main():
+  get_certificates()
+  run_post()
+  #TODO: create a dummy certificate, if no certificate exists
+  while True:
+    now = datetime.datetime.now()
+    renew_time = now + datetime.timedelta(days=1)
+    renew_time = renew_time.replace(hour=3, minute=0, second=0, microsecond=0)
+    time.sleep((renew_time - now).total_seconds())
+    run_certbot_renew()
+
+if len(sys.argv) > 1:
+  run_post()
+else:
+  run_main()
```

Comment:
(this file) :+1:

Reply:
I'd love the whole setup to be ready-to-run for sysadmins wanting to contribute. That means: no letsencrypt in development. What about running certbot if the relevant (or a specific) environment variable is set (e.g. `CERTBOT_EMAIL` - which would need to move out of `docker-compose.yml`, `CERTBOT_ENABLE` or `USE_CERTBOT`), and generating the test certificate if not?

Since haproxy is pretty smart about selecting the right certificate, I'm pondering whether it would be useful to _always_ generate the test certificate - so that at least haproxy starts if certbot fails (e.g. over limit) and one can reach the apps by bypassing security warnings. That may be better than being down completely (can't check if the rest works, for example).

Reply:
there is already a `#TODO: create a dummy certificate` in the script. I'd generate a `localhost` certificate if there are no other certificate and remove it when the first certificate has been generated. so it should be possible to get `haproxy` running